cybench/models/residual_models.py

```diff
@@ -0,0 +1,158 @@
+import numpy as np
+import pandas as pd
+import pickle
+
+from cybench.datasets.dataset import Dataset
+from cybench.datasets.modified_dataset import ModifiedTargetsDataset
+from cybench.models.model import BaseModel
+from cybench.models.trend_models import TrendModel
+from cybench.models.sklearn_models import SklearnRidge, SklearnRandomForest
+from cybench.models.nn_models import BaselineLSTM, BaselineInceptionTime
+from cybench.util.data import data_to_pandas
+
+from cybench.config import (
+    KEY_LOC,
+    KEY_YEAR,
+    KEY_TARGET,
+)
+
+
+class ResidualModel(BaseModel):
+    def __init__(self, baseline_model: BaseModel):
+        """Ridge model that predicts residuals from the trend.
+
+        It leverages `TrendModel` to estimate the trend.
+        Then it detrends targets and predicts the residuals.
+        During inference, residual predictions are added to trend to
+        produce the target predictions.
+        """
+        self._trend_model = TrendModel()
+        self._baseline_model = baseline_model
+
+    def fit(self, dataset: Dataset, **fit_params):
+        """Fit or train the model.
+
+        Args:
+          train_dataset (Dataset): training dataset
+          **fit_params: Additional parameters.
+
+        Returns:
+          A tuple containing the fitted model and a dict with additional information.
+        """
+        train_df = data_to_pandas(dataset, data_cols=[KEY_LOC, KEY_YEAR, KEY_TARGET])
+
+        train_residuals = []
+        for item in dataset:
+            trend_train_df = train_df[
+                (train_df[KEY_LOC] == item[KEY_LOC])
+                & (train_df[KEY_YEAR] != item[KEY_YEAR])
+            ].copy()
+            # only one year of training data for this location
+            if trend_train_df.empty:
+                trend_pred = train_df[KEY_TARGET].mean()
+            else:
+                trend_train_df.set_index([KEY_LOC, KEY_YEAR], inplace=True)
+                trend_train_dataset = Dataset(dataset.crop, data_target=trend_train_df)
+                self._trend_model.fit(trend_train_dataset)
+                trend_pred, _ = self._trend_model.predict_items([item])
+                if not isinstance(trend_pred, float):
+                    trend_pred = trend_pred[0]
+
+            res_row = {
+                KEY_LOC: item[KEY_LOC],
+                KEY_YEAR: item[KEY_YEAR],
+                KEY_TARGET: trend_pred - item[KEY_TARGET],
+            }
+            train_residuals.append(res_row)
+
+        residuals_df = pd.DataFrame(train_residuals)
+        residuals_df.set_index([KEY_LOC, KEY_YEAR], inplace=True)
+        residuals_dataset = ModifiedTargetsDataset(
+            dataset, modified_targets=residuals_df
+        )
+        self._baseline_model.fit(residuals_dataset, **fit_params)
+
+        # Fit the trend model on the entire training data
```

Comment:
Thi may need a bit of extra explanation; why for computing the residual the year at hand is excluded, while here it is included

Reply:
Above we are predicting for training data itself. One item at a time. Excluding the item for which we are predicting. This is something that could be skipped. We could fit on the entire training data there also.